xbmc/addons/AddonRepos.cpp

```diff
@@ -0,0 +1,220 @@
+#include "CompileInfo.h"
+#include "utils/StringUtils.h"
+#include "utils/log.h"
+
+#include <vector>
+
+using namespace ADDON;
+
+/**********************************************************
+ * CAddonRepos
+ *
+ */
+
+namespace
+{
+
+std::vector<RepoInfo> LoadOfficialRepoInfos()
+{
+  const std::vector<std::string> officialAddonRepos =
+      StringUtils::Split(CCompileInfo::GetOfficialAddonRepos(), ',');
+
+  std::vector<RepoInfo> officialRepoInfos;
+  RepoInfo newRepoInfo;
+
+  for (const auto& addonRepo : officialAddonRepos)
+  {
+    const std::vector<std::string> tmpRepoInfo = StringUtils::Split(addonRepo, '|');
+    newRepoInfo.m_repoId = tmpRepoInfo.front();
+    newRepoInfo.m_origin = tmpRepoInfo.back();
+    officialRepoInfos.emplace_back(newRepoInfo);
+  }
+
+  return officialRepoInfos;
+}
+
+} // unnamed namespace
+
+static std::vector<RepoInfo> officialRepoInfos = LoadOfficialRepoInfos();
+
+bool CAddonRepos::IsFromOfficialRepo(const AddonPtr& addon)
+{
+  return IsFromOfficialRepo(addon, false);
+}
+
+bool CAddonRepos::IsFromOfficialRepo(const AddonPtr& addon, bool bCheckAddonPath)
+{
+  auto comparator = [&](const RepoInfo& officialRepo) {
+    if (bCheckAddonPath)
+    {
+      return (addon->Origin() == officialRepo.m_repoId &&
+              StringUtils::StartsWithNoCase(addon->Path(), officialRepo.m_origin));
+    }
+
+    return addon->Origin() == officialRepo.m_repoId;
+  };
+
+  return addon->Origin() == ORIGIN_SYSTEM ||
+         std::any_of(officialRepoInfos.begin(), officialRepoInfos.end(), comparator);
+}
+
+const std::map<std::string, AddonPtr>& CAddonRepos::GetLatestOfficialVersions() const
+{
+  return m_latestOfficialVersions;
+}
+
+const std::map<std::string, std::map<std::string, AddonPtr>>& CAddonRepos::GetLatestVersionsByRepo()
+    const
+{
+  return m_latestVersionsByRepo;
+}
+
+void CAddonRepos::SetupFullAddonsByRepoMap(const VECADDONS& addons)
+{
+  m_addonsByRepoMap.clear();
+
+  for (const auto& addon : addons)
+    if (m_addonMgr.IsCompatible(*addon))
+    {
+      m_addonsByRepoMap[addon->Origin()].insert({addon->ID(), addon});
+    }
+
+  for (const auto& map : m_addonsByRepoMap)
+    CLog::Log(LOGDEBUG, "Setting up map for repository: {} - addon count : {}", map.first,
+              map.second.size());
+
+  SetupLatestVersionMaps();
+}
+
+void CAddonRepos::SetupLatestVersionMaps()
+{
+  m_latestOfficialVersions.clear();
+  m_latestVersionsByRepo.clear();
+
+  const AddonRepoUpdateMode updateMode =
+      CAddonSystemSettings::GetInstance().GetAddonRepoUpdateMode();
+
+  for (const auto& repo : m_addonsByRepoMap)
+  {
+    const auto& addonsPerRepo = repo.second;
+
+    for (const auto& addonMapEntry : addonsPerRepo)
+    {
+      const auto& addonToAdd = addonMapEntry.second;
+
+      if (updateMode == AddonRepoUpdateMode::OFFICIAL_ONLY)
+      {
+        if (IsFromOfficialRepo(addonToAdd, true))
+        {
+          AddAddonIfLatest(addonToAdd, m_latestOfficialVersions);
+        }
```

Comment:
should this be:

```cpp
else
{
      // add to latestVersionsByRepo - map
      AddAddonIfLatest(repo.first, addonToAdd, m_latestVersionsByRepo);
}
```

Reply:
Any reason we would use the latestVersionsByRepo in this mode?

Reply:
I don't think there should be any unofficial addons in `m_latestVersionsByRepo`.

Reply:
it's getting populated a few lines later, or am i missing something here?
https://github.com/xbmc/xbmc/blob/e54f145923f00b28f409642d20c34051ed499b98/xbmc/addons/AddonRepos.cpp#L130-L131

Reply:
...but it's not used outside of official-mode, right

Reply:
fixed